hoss/dimension_utilities.py

```diff
@@ -75,8 +75,143 @@ def prefetch_dimension_variables(opendap_url: str, varinfo: VarInfoFromDmr,
 
     logger.info('Variables being retrieved in prefetch request: '
                 f'{format_variable_set_string(required_dimensions)}')
-    return get_opendap_nc4(opendap_url, required_dimensions, output_dir,
-                           logger, access_token, config)
+    
+    required_dimensions_nc4 = get_opendap_nc4(opendap_url,
+                                              required_dimensions, output_dir,logger, access_token, config)
+    
+    # Create bounds variables if necessary. 
+    add_bounds_variables(required_dimensions_nc4, required_dimensions,
+                         varinfo, logger)
+    
+    return required_dimensions_nc4
+    
+
+def add_bounds_variables(dimensions_nc4: str,
+                         required_dimensions: Set[str],
+                         varinfo: VarInfoFromDmr,
+                         logger: Logger) -> None:
+    """ Augment a NetCDF4 file with artificial bounds variables for each 
+        dimension variable that is edge-aligned and does not already 
+        have bounds variables. 
+
+        For each dimension variable:
+        (1) Check if the variable needs a bounds variable.
+        (2) If so, create a bounds array of minimum and maximum values.
+        (3) Then write the bounds variable to the NetCDF4 URL.
+
+    """
+    with Dataset(dimensions_nc4, 'r+') as datasets:
+        for dimension_name in required_dimensions:
+            dimension_variable = varinfo.get_variable(dimension_name)
+            if needs_bounds(dimension_variable) is True:
+                min_and_max_bounds = create_bounds(datasets, dimension_name)
+                write_bounds(datasets, dimension_variable, min_and_max_bounds)
+
+                logger.info('Artificial bounds added for dimension variable: '
+                f'{dimension_name}')
+
+        
+def needs_bounds(dimension: VariableFromDmr) -> bool:
+    """ Check if a dimension variable needs a bounds variable.
+        This will be the case when dimension cells are edge-aligned 
+        and bounds for that dimension do not already exist.
+    
+    """
+    
+    return dimension.attributes['cell_alignment'] == 'edge' and dimension.references.get('bounds') == None
```

Comment:
I took a look at the [results from the tests](https://github.com/nasa/harmony-opendap-subsetter/actions/runs/8061068393/job/22018147013), here are the styling nits `pycodestyle` found:

```
************* Module hoss.dimension_utilities
hoss/dimension_utilities.py:78:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:80:0: C0301: Line too long (107/100) (line-too-long)
hoss/dimension_utilities.py:81:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:82:43: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:85:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:87:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:113:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:120:0: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:121:0: C0301: Line too long (106/100) (line-too-long)
hoss/dimension_utilities.py:152:69: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:158:43: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:159:45: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:164:50: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:188:79: C0303: Trailing whitespace (trailing-whitespace)
hoss/dimension_utilities.py:194:0: C0301: Line too long (104/100) (line-too-long)
hoss/dimension_utilities.py:121:64: C0121: Comparison 'dimension.references.get('bounds') == None' should be 'dimension.references.get('bounds') is None' (singleton-comparison)
```

There are also some failed tests, but I suspect at least some of them might stem from the `cell_alignment` comment above.

Reply:
We shouldn't assume that dimension.attributes has a cell_alignment key. To be safe against a lack of it, it would be better to use the dict.get method, e.g.: dimension.attributes.get('cell_alignment').
Oh, so will the `attributes.get()` be able to return `None` whereas my original direct call to `attributes` would flounder if a 'cell_alignment' attribute didn't exist?

Reply:
That's exactly it, yeah.

Reply:
@flamingbear I did not know that was an option. It's turned on now!